### PR TITLE
Adds a better way to include binary resources

### PR DESCRIPTION
**Pull request by bnaecker**
- Uses a small assembly file to include binary resources, rather than
  objcopy
- Updates `base/common/getResource.cpp` for this new method of inclusion
- Removes linux-only guards in CMake files, as this solution is
  cross-platform.

The resulting binary resources are available in the ClickHouse server
binary on Linux, macOS, and illumos platforms. FreeBSD has not been
tested, but will likely work as well.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Adds cross-platform embedding of binary resources into executables.

Detailed description / Documentation draft:

This commit changes how binary resource files are embedded into object files. The previous solution relied on `objcopy`, which implies a lot of platform-specific behavior, architecture detection, and an external tool. This commit instead uses the `.incbin` assembly directive to include a file directly from its path during compilation. This is more portable and does not rely on an external tool, only the C compiler driver. As a result, the timezone data and configuration files can be embedded into the server binary for all supported platforms, not only Linux.

### CODE REVIEW

File: base/common/getResource.cpp [hunk 1]
```diff
@@ -4,23 +4,41 @@
...
+    std::string symbol_name_end = prefix + name_replaced + "_end";
 
-    const void * sym_data = dlsym(RTLD_DEFAULT, symbol_name_data.c_str());
-    const void * sym_size = dlsym(RTLD_DEFAULT, symbol_name_size.c_str());
+    auto sym_start = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_start.c_str()));
+    auto sym_end = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_end.c_str()));
 
-    if (sym_data && sym_size)
-        return { static_cast<const char *>(sym_data), unalignedLoad<size_t>(&sym_size) };
+    if (sym_start && sym_end) {
```
Comment (alexey-milovidov) [inline base/common/getResource.cpp:39]: Style check will argue about `{` on the same line (please move to the next line).

**Review by alexey-milovidov — APPROVED** [review]

File: base/common/getResource.cpp (hunk 1 above)
Comment (bnaecker) [reply base/common/getResource.cpp:39]: Will do, thanks.

**Comment by alexey-milovidov** [comment]
AArch64 build:
> Jun 09 23:42:47 ld.lld: error: /build/build_docker/contrib/cctz-cmake/libtzdata.a(Abidjan.o) is incompatible with elf64-littleaarch64

**Comment by alexey-milovidov** [comment]
clang-tidy:
```
2021-06-09 23:42:23 /build/obj-x86_64-linux-gnu/../base/common/getResource.cpp:36:5: error: 'auto sym_start' can be declared as 'const auto *sym_start' [readability-qualified-auto,-warnings-as-errors]
2021-06-09 23:42:23     auto sym_start = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_start.c_str()));
2021-06-09 23:42:23     ^~~~~
2021-06-09 23:42:23     const auto *
2021-06-09 23:42:23 /build/obj-x86_64-linux-gnu/../base/common/getResource.cpp:37:5: error: 'auto sym_end' can be declared as 'const auto *sym_end' [readability-qualified-auto,-warnings-as-errors]
2021-06-09 23:42:23     auto sym_end = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_end.c_str()));
2021-06-09 23:42:23     ^~~~~
2021-06-09 23:42:23     const auto *
```

**Comment by alexey-milovidov** [comment]
PS. Just writing `const char *` is fine.

**Comment by bnaecker** [comment]
> clang-tidy:
> 
> ```
> 2021-06-09 23:42:23 /build/obj-x86_64-linux-gnu/../base/common/getResource.cpp:36:5: error: 'auto sym_start' can be declared as 'const auto *sym_start' [readability-qualified-auto,-warnings-as-errors]
> 2021-06-09 23:42:23     auto sym_start = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_start.c_str()));
> 2021-06-09 23:42:23     ^~~~~
> 2021-06-09 23:42:23     const auto *
> 2021-06-09 23:42:23 /build/obj-x86_64-linux-gnu/../base/common/getResource.cpp:37:5: error: 'auto sym_end' can be declared as 'const auto *sym_end' [readability-qualified-auto,-warnings-as-errors]
> 2021-06-09 23:42:23     auto sym_end = reinterpret_cast<const char*>(dlsym(RTLD_DEFAULT, symbol_name_end.c_str()));
> 2021-06-09 23:42:23     ^~~~~
> 2021-06-09 23:42:23     const auto *
> ```

This is resolved, thanks.

**Comment by bnaecker** [comment]
> AArch64 build:
> 
> > Jun 09 23:42:47 ld.lld: error: /build/build_docker/contrib/cctz-cmake/libtzdata.a(Abidjan.o) is incompatible with elf64-littleaarch64

Is there any more detail or context for this error? I'm not sure which GitHub check (if any) that comes from.

File: CMakeLists.txt [hunk 2]
```diff
@@ -183,24 +183,31 @@ endif ()
...
 
-if (OS_LINUX)
-    find_program (OBJCOPY_PATH NAMES "llvm-objcopy" "llvm-objcopy-12" "llvm-objcopy-11" "llvm-objcopy-10" "llvm-objcopy-9" "llvm-objcopy-8" "objcopy")
-    if (OBJCOPY_PATH)
-        message(STATUS "Using objcopy: ${OBJCOPY_PATH}.")
-
+find_program (OBJCOPY_PATH NAMES "llvm-objcopy" "llvm-objcopy-12" "llvm-objcopy-11" "llvm-objcopy-10" "llvm-objcopy-9" "llvm-objcopy-8" "objcopy")
+if (OBJCOPY_PATH)
+   message(STATUS "Using objcopy: ${OBJCOPY_PATH}.")
+   if (OS_LINUX)
```
Comment (bnaecker) [inline CMakeLists.txt:189]: I left this chunk setting `OBJCOPY_ARCH_OPTIONS`. It's no longer used as far as I can tell, but I wasn't sure if there were plans to use it later. Let me know if you'd like to remove this section in this PR as well.

Comment (alexey-milovidov) [reply CMakeLists.txt:189]: We have no plans to reuse this option, you can remove it freely.

**Comment by alexey-milovidov** [comment]
> Is there any more detail or context for this error? I'm not sure which GitHub check (if any) that comes from.

`ClickHouse special build check — 3/5 builds are OK`, `clang-11-aarch64`, `Build log`.

**Comment by bnaecker** [comment]
> > Is there any more detail or context for this error? I'm not sure which GitHub check (if any) that comes from.
> 
> `ClickHouse special build check — 3/5 builds are OK`, `clang-11-aarch64`, `Build log`.

Thanks, I see it now. However, I'm not able to run the Docker build container as that action describes. Running exactly the command in the first line of the error message downloads the image and starts a container from it, but soon errors with:

```bash
bnaecker@granite:~/clickhouse$ docker run --network=host --volume=/place/sandbox-data/ramdrive/991451814/ClickHouse/build_output_folder:/output --volume=/place/sandbox-data/ramdrive/991451814/ClickHouse/docker/packager/../..:/build --volume=/place/sandbox-data/ramdrive/991451814/local_ccache:/ccache  -e CC=clang-11 -e CXX=clang++-11 -e CCACHE_DIR=/ccache -e CCACHE_BASEDIR=/build -e CCACHE_NOHASHDIR=true -e CCACHE_COMPILERCHECK=content -e CCACHE_MAXSIZE=15G -e VERSION_STRING='21.7.1.7110' -e AUTHOR='clickhouse' -e CMAKE_FLAGS="$CMAKE_FLAGS -DADD_GDB_INDEX_FOR_GOLD=1 -DCMAKE_TOOLCHAIN_FILE=/build/cmake/linux/toolchain-aarch64.cmake -DCMAKE_C_COMPILER=`which clang-11` -DCMAKE_CXX_COMPILER=`which clang++-11`"  yandex/clickhouse-binary-builder:latest
Jun 11 21:57:42 + mkdir -p build/cmake/toolchain/darwin-x86_64
Jun 11 21:57:42 + tar xJf MacOSX10.15.sdk.tar.xz -C build/cmake/toolchain/darwin-x86_64 --strip-components=1
Jun 11 21:57:43 + mkdir -p build/cmake/toolchain/linux-aarch64
Jun 11 21:57:43 + tar xJf gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu.tar.xz -C build/cmake/toolchain/linux-aarch64 --strip-components=1
Jun 11 21:57:59 + mkdir -p build/cmake/toolchain/freebsd-x86_64
Jun 11 21:57:59 + tar xJf freebsd-11.3-toolchain.tar.xz -C build/cmake/toolchain/freebsd-x86_64 --strip-components=1
Jun 11 21:58:00 + mkdir -p build/build_docker
Jun 11 21:58:00 + cd build/build_docker
Jun 11 21:58:00 + rm -f CMakeCache.txt
Jun 11 21:58:00 + read -ra CMAKE_FLAGS
Jun 11 21:58:00 + cmake --debug-trycompile --verbose=1 -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE= -DSANITIZE= -DENABLE_CHECK_HEAVY_BUILDS=1 -DADD_GDB_INDEX_FOR_GOLD=1 -DCMAKE_TOOLCHAIN_FILE=/build/cmake/linux/toolchain-aarch64.cmake -DCMAKE_C_COMPILER= -DCMAKE_CXX_COMPILER= ..
Jun 11 21:58:00 debug trycompile on
Jun 11 21:58:00 CMake Error: The source directory "/build/build_docker/--verbose=1" does not exist.
Jun 11 21:58:00 Specify --help for usage, or press the help button on the CMake GUI.
Jun 11 21:58:00 -- Cache values
```

Please let me know if there's some other setup required to get the image to launch a container correctly. I'm also trying to build locally with the options that seems to require (Clang 11, LLVM linker, AArch64 target), we'll see how that works.

As an aside, it seems this error arises because we're cross compiling. Presumably this is being build on an x86 system, but we're generating code for ARM. Is that correct? The new version of the macro here just calls into the C compiler directly, so I'm just trying to understand how the target architecture is being set by the build scripts.

**Comment by alexey-milovidov** [comment]
> As an aside, it seems this error arises because we're cross compiling. Presumably this is being build on an x86 system, but we're generating code for ARM. Is that correct?

Yes, that's correct.

> I'm just trying to understand how the target architecture is being set by the build scripts.

It is being set in `cmake/linux/toolchain-aarch64.cmake`

**Comment by alexey-milovidov** [comment]
> `docker run --network=host --volume=/place/sandbox-data/ramdrive/991451814/ClickHouse/build_output_folder:/output --volume=/place/sandbox-data/ramdrive/991451814/ClickHouse/docker/packager/../..:/build --volume=/place/sandbox-data/ramdrive/991451814/local_ccache:/ccache  -e CC=clang-11 -e CXX=clang++-11 -e CCACHE_DIR=/ccache -e CCACHE_BASEDIR=/build -e CCACHE_NOHASHDIR=true -e CCACHE_COMPILERCHECK=content -e CCACHE_MAXSIZE=15G -e VERSION_STRING='21.7.1.7110' -e AUTHOR='clickhouse' -e CMAKE_FLAGS="$CMAKE_FLAGS -DADD_GDB_INDEX_FOR_GOLD=1 -DCMAKE_TOOLCHAIN_FILE=/build/cmake/linux/toolchain-aarch64.cmake -DCMAKE_C_COMPILER=`which clang-11` -DCMAKE_CXX_COMPILER=`which clang++-11`"  yandex/clickhouse-binary-builder:latest`

The ccache volume can be removed and also you have to adjust the path (`/place/sandbox-data/...`)

**Comment by bnaecker** [comment]
I've now verified this works when cross-compiling from x86 to ARM.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**